File: npc_sim/npc/npc.py
# Copyright 2025-2026 Sadık Abdusselam Albayrak
# Licensed under the Apache License, Version 2.0
"""Primary NPC data container aggregating all subsystems."""

from __future__ import annotations
import math
from npc_sim.core.sim_vector3 import SimVector3
from npc_sim.core.sim_config import SimulationConfig
from npc_sim.events.sim_event import SimEvent
from npc_sim.npc.identity import NPCIdentity
from npc_sim.npc.vitals import NPCVitals
from npc_sim.npc.psychology import NPCPsychology
from npc_sim.npc.social import NPCSocial
from npc_sim.npc.memory import NPCMemory
from npc_sim.npc.beliefs import BeliefSystem
from npc_sim.npc.goals import NPCGoals, Goal, GoalType
from npc_sim.npc.traits import NPCTraits
from npc_sim.npc.inventory import NPCInventory
from npc_sim.npc.schedule import NPCSchedule


class NPC:
    """
    The primary NPC data container.
    Aggregates Vitals, Identity, Psychology, Social, Belief, Memory, Goals, Traits, Inventory.
    All logic lives in Systems — this class is pure data + minimal wiring.
    """

    def __init__(
        self,
        identity: NPCIdentity,
        vitals: NPCVitals = None,
        psychology: NPCPsychology = None,
        social: NPCSocial = None,
        memory_capacity: int = 50,
        traits: NPCTraits = None,
        inventory_slots: int = 10,
    ):
        if identity is None:
            raise ValueError("identity is required")

        self.identity = identity
        self.vitals = vitals or NPCVitals()
        self.psychology = psychology or NPCPsychology()
        self.social = social or NPCSocial()
        self.beliefs = BeliefSystem()
        self.memory = NPCMemory(memory_capacity)
        self.goals = NPCGoals()
        self.traits = traits or NPCTraits()
        self.inventory = NPCInventory(inventory_slots)
        self.schedule = NPCSchedule.for_occupation(identity.occupation)

        # World state
        self.position = SimVector3.ZERO
        self.forward = SimVector3(0, 0, 1)
        self.is_active = True

        # Config reference (injected by SimulationManager)
        self._config: SimulationConfig | None = None

    def inject_config(self, config: SimulationConfig) -> None:
        self._config = config

    def set_schedule(self, schedule: NPCSchedule) -> None:
        if schedule is not None:
            self.schedule = schedule

    # ── Tick ──

    def tick(self, delta_time: float, current_time: float) -> None:
        if not self.is_active or not self.vitals.is_alive:
            return

        cfg = self._config
        hunger_rate = cfg.hunger_decay_rate if cfg else 0.001
        thirst_rate = cfg.thirst_decay_rate if cfg else 0.002
        energy_rate = cfg.energy_decay_rate if cfg else 0.5
        mem_decay = cfg.global_memory_decay_rate if cfg else 0.001
        belief_decay = cfg.global_belief_decay_rate if cfg else 0.005
        rel_decay = cfg.relation_decay_rate if cfg else 0.00005
        fear_rate = cfg.fear_decay_rate if cfg else 0.0005
        happy_rate = cfg.happiness_decay_rate if cfg else 0.0003
        anger_rate = cfg.anger_decay_rate if cfg else 0.0004

        # Physiology
        self.vitals.set_hunger(self.vitals.hunger + hunger_rate * delta_time)
        self.vitals.set_thirst(self.vitals.thirst + thirst_rate * delta_time)
        self.vitals.consume_energy(energy_rate * delta_time)

        # Death by neglect
        if cfg and cfg.death_by_neglect:
            if self.vitals.hunger >= 1.0 or self.vitals.thirst >= 1.0:
                self.vitals.apply_damage(10.0 * delta_time)

        # Cross-component: Stress → Anger (Neuroticism-scaled)
        stress_anger = self.vitals.stress * 0.02 * delta_time * self.psychology.neuroticism
        self.psychology.set_anger(self.psychology.anger + stress_anger)

        # Hunger/Thirst → Stress
        need_stress = (self.vitals.hunger + self.vitals.thirst) * 0.01 * delta_time
        self.vitals.set_stress(self.vitals.stress + need_stress)

        # Emotional decay
        self.psychology.decay_emotions(delta_time, fear_rate, happy_rate, anger_rate)

        # Relation decay
        self.social.tick_decay(delta_time, rel_decay)

        # Memory and belief decay
        self.memory.decay_all(mem_decay * delta_time)
        self.beliefs.decay_all(belief_decay * delta_time)

        # Prune expired goals
        self.goals.prune_expired(current_time)

    # ── Event witnessing ──

    def witness_event(self, sim_event: SimEvent, belief_subjects: list[str], current_time: float) -> None:
        if sim_event is None:
            return

        # Record in episodic memory
        emotional_weight = sim_event.impact * (1.0 - self.psychology.neuroticism * 0.5)
        self.memory.remember(sim_event, emotional_weight, current_time)

        # Update beliefs about each subject
        self.beliefs.process_event(sim_event, belief_subjects, current_time)

        # Emotional reaction
        stress_delta = abs(sim_event.impact) * 0.1 * self.psychology.neuroticism
        self.vitals.set_stress(self.vitals.stress + stress_delta)

        if sim_event.impact < -0.3:
            fear_spike = abs(sim_event.impact) * 0.15 * self.psychology.neuroticism
            self.psychology.set_fear(self.psychology.fear + fear_spike)
        elif sim_event.impact > 0.2:
            happy_boost = sim_event.impact * 0.1 * (1.0 - self.psychology.neuroticism * 0.5)
            self.psychology.set_happiness(self.psychology.happiness + happy_boost)

    # ── Social interaction ──

    def interact(self, other: NPC, trust_delta: float, affinity_delta: float,
                 respect_delta: float, current_time: float) -> None:
        if other is None:
            return
        relation = self.social.get_or_create_relation(self.identity.npc_id, other.identity.npc_id)
        relation.apply_interaction(trust_delta, affinity_delta, respect_delta, current_time)

    # ── Need → Goal pipeline ──

    def refresh_need_goals(self, current_time: float, urgency_window: float = 300.0) -> None:
        expiry = current_time + urgency_window

        if self.vitals.hunger > 0.65 and not self.goals.has_active_goal(GoalType.FIND_FOOD):
            self.goals.add_goal(Goal(GoalType.FIND_FOOD, "I need to eat", self.vitals.hunger, expiry))

        if self.vitals.thirst > 0.70 and not self.goals.has_active_goal(GoalType.FIND_WATER):
            self.goals.add_goal(Goal(GoalType.FIND_WATER, "I need water", self.vitals.thirst, expiry))

        if self.vitals.energy < 0.30 * self.vitals.max_energy and not self.goals.has_active_goal(GoalType.REST):
            self.goals.add_goal(Goal(GoalType.REST, "I need rest", 1.0 - self.vitals.energy_norm, expiry))

    # ── Lifecycle ──

    def deactivate(self) -> None:
        self.is_active = False

    def to_dict(self) -> dict:
        return {
            "identity": self.identity.to_dict(),
            "vitals": self.vitals.to_dict(),
            "psychology": self.psychology.to_dict(),
            "social": self.social.to_dict(),
            "inventory": self.inventory.to_dict(),
            "traits": self.traits.to_dict(),
            "position": {"x": round(self.position.x, 2), "y": round(self.position.y, 2), "z": round(self.position.z, 2)},
            "is_active": self.is_active,
            "schedule": self.schedule.to_dict(),
        }

    def __repr__(self) -> str:
        state = "Active" if self.is_active else "Inactive"
        return f"[NPC] {self.identity.display_name} ({state}) | {self.vitals} | {self.psychology}"

File: Runtime/NPC/NPC.cs
using System;
using System.Collections.Generic;

namespace ForgeProject.Sim
{
    /// <summary>
    /// The primary NPC data container.
    /// Aggregates Vitals, Identity, Psychology, Social, Belief, Memory, Goals, Traits, and Inventory.
    /// All logic lives in Systems — this class is pure data + minimal wiring.
    /// </summary>
    public class NPC
    {
        // ─── Components ───────────────────────────────────────────────────────────
        public NPCVitals     Vitals      { get; private set; }
        public NPCIdentity   Identity    { get; private set; }
        public NPCPsychology Psychology  { get; private set; }
        public NPCSocial     Social      { get; private set; }
        public BeliefSystem  Beliefs     { get; private set; }
        public NPCMemory     Memory      { get; private set; }
        public NPCGoals      Goals       { get; private set; }
        public NPCTraits     Traits      { get; private set; }
        public NPCInventory  Inventory   { get; private set; }
        public NPCSchedule   Schedule    { get; private set; }

        // ─── World State ──────────────────────────────────────────────────────────
        /// <summary>Current world position. Updated by movement actions / SimWorldAdapter.</summary>
        public SimVector3 Position { get; set; } = SimVector3.Zero;

        /// <summary>Forward facing direction (used by SensorRange FOV check).</summary>
        public SimVector3 Forward  { get; set; } = new SimVector3(0, 0, 1);

        /// <summary>Whether this NPC is participating in the active simulation.</summary>
        public bool IsActive { get; private set; } = true;

        // ─── Config reference (injected by SimulationManager) ─────────────────────
        private SimulationConfig _config;

        // ─── Constructor ──────────────────────────────────────────────────────────
        /// <summary>Creates a fully configured NPC. Config is injected later by SimulationManager.</summary>
        public NPC(NPCIdentity    identity,
                   NPCVitals      vitals        = null,
                   NPCPsychology  psychology    = null,
                   NPCSocial      social        = null,
                   int            memoryCapacity = 50,
                   NPCTraits      traits        = null,
                   int            inventorySlots = 10)
        {
            Identity   = identity   ?? throw new ArgumentNullException(nameof(identity));
            Vitals     = vitals     ?? new NPCVitals();
            Psychology = psychology ?? new NPCPsychology();
            Social     = social     ?? new NPCSocial();
            Beliefs    = new BeliefSystem();
            Memory     = new NPCMemory(memoryCapacity);
            Goals      = new NPCGoals();
            Traits     = traits     ?? new NPCTraits();
            Inventory  = new NPCInventory(inventorySlots);
            Schedule   = NPCSchedule.ForOccupation(identity.Occupation);
        }

        // ─── Config injection ─────────────────────────────────────────────────────
        /// <summary>Called by SimulationManager.AddNPC to wire simulation-wide config rates.</summary>
        public void InjectConfig(SimulationConfig config) => _config = config;

        /// <summary>
        /// Replaces this NPC's daily schedule. Called by the Bridge layer when a
        /// <see cref="ForgeProject.Sim.Bridge.NPCScheduleSO"/> override is assigned.
        /// </summary>
        public void SetSchedule(NPCSchedule schedule) => Schedule = schedule ?? Schedule;

        // ─── Tick ─────────────────────────────────────────────────────────────────
        /// <summary>
        /// Called every simulation step via <see cref="SimulationManager"/>.
        /// Applies physiological decay, emotional decay, relation decay, memory/belief decay,
        /// goal expiry pruning, and cross-component stress links.
        /// </summary>
        public void Tick(float deltaTime, float currentTime)
        {
            if (!IsActive || !Vitals.IsAlive) return;

            // Use config rates where available, fallback to sensible defaults
            float hungerRate  = _config?.HungerDecayRate         ?? 0.001f;
            float thirstRate  = _config?.ThirstDecayRate         ?? 0.002f;
            float energyRate  = _config?.EnergyDecayRate         ?? 0.5f;
            float memDecay    = _config?.GlobalMemoryDecayRate   ?? 0.001f;
            float beliefDecay = _config?.GlobalBeliefDecayRate   ?? 0.005f;
            float relDecay    = _config?.RelationDecayRate       ?? 0.00005f;
            float fearRate    = _config?.FearDecayRate           ?? 0.0005f;
            float happyRate   = _config?.HappinessDecayRate      ?? 0.0003f;
            float angerRate   = _config?.AngerDecayRate          ?? 0.0004f;

            // ── Physiology (D1 fix) ──
            Vitals.SetHunger(Vitals.Hunger + hungerRate * deltaTime);
            Vitals.SetThirst(Vitals.Thirst + thirstRate * deltaTime);
            Vitals.ConsumeEnergy(energyRate * deltaTime);

            // ── DeathByNeglect ──
            if (_config?.DeathByNeglect == true)
            {
                if (Vitals.Hunger >= 1f || Vitals.Thirst >= 1f)
                    Vitals.ApplyDamage(10f * deltaTime);
            }

            // ── Cross-component: Stress → Anger (Neuroticism-scaled) ──
            float stressAngerSpill = Vitals.Stress * 0.02f * deltaTime * Psychology.Neuroticism;
            Psychology.SetAnger(Psychology.Anger + stressAngerSpill);

            // ── Hunger/Thirst → Stress ──
            float needStress = (Vitals.Hunger + Vitals.Thirst) * 0.01f * deltaTime;
            Vitals.SetStress(Vitals.Stress + needStress);

            // ── Emotional decay (D6 fix) ──
            Psychology.DecayEmotions(deltaTime, fearRate, happyRate, angerRate);

            // ── Relation decay (D4 fix) ──
            Social.TickDecay(deltaTime, relDecay);

            // ── Memory and belief decay ──
            Memory.DecayAll(memDecay * deltaTime);
            Beliefs.DecayAll(beliefDecay * deltaTime);

            // ── Prune expired goals ──
            Goals.PruneExpired(currentTime);
        }

        // ─── Event witnessing ─────────────────────────────────────────────────────
        /// <summary>
        /// NPC witnesses a SimEvent — stores it in memory, updates beliefs, and reacts emotionally.
        /// </summary>
        public void WitnessEvent(SimEvent simEvent, IEnumerable<string> beliefSubjects, float currentTime)
        {
            if (simEvent == null) return;

            // Record in episodic memory
            float emotionalWeight = simEvent.Impact * (1f - Psychology.Neuroticism * 0.5f);
            Memory.Remember(simEvent, emotionalWeight, currentTime);

            // Update beliefs about each subject
            Beliefs.ProcessEvent(simEvent, beliefSubjects, currentTime);

            // Emotional reaction (D6 fix) ─────────────────────────────────────────
            float stressDelta = MathF.Abs(simEvent.Impact) * 0.1f * Psychology.Neuroticism;
            Vitals.SetStress(Vitals.Stress + stressDelta);

            if (simEvent.Impact < -0.3f)
            {
                // Negative event → Fear spike (Neuroticism-amplified)
                float fearSpike = MathF.Abs(simEvent.Impact) * 0.15f * Psychology.Neuroticism;
                Psychology.SetFear(Psychology.Fear + fearSpike);
            }
            else if (simEvent.Impact > 0.2f)
            {
                // Positive event → Happiness boost (dampened by Neuroticism)
                float happyBoost = simEvent.Impact * 0.1f * (1f - Psychology.Neuroticism * 0.5f);
                Psychology.SetHappiness(Psychology.Happiness + happyBoost);
            }
        }

        // ─── Social interaction ───────────────────────────────────────────────────
        /// <summary>Interacts with another NPC, updating the relation between them.</summary>
        public void Interact(NPC other, float trustDelta, float affinityDelta, float respectDelta, float currentTime)
        {
            if (other == null) return;
            var relation = Social.GetOrCreateRelation(Identity.NpcId, other.Identity.NpcId);
            relation.ApplyInteraction(trustDelta, affinityDelta, respectDelta, currentTime);
        }

        // ─── Need → Goal pipeline ─────────────────────────────────────────────────
        /// <summary>
        /// Converts urgent physiological needs into active goals.
        /// Called by NeedSystem each tick — avoids goal flooding by checking HasActiveGoal.
        /// </summary>
        public void RefreshNeedGoals(float currentTime, float urgencyWindow = 300f)
        {
            float expiry = currentTime + urgencyWindow;

            if (Vitals.Hunger > 0.65f && !Goals.HasActiveGoal(GoalType.FindFood))
                Goals.AddGoal(new Goal(GoalType.FindFood, "I need to eat", Vitals.Hunger, expiry));

            if (Vitals.Thirst > 0.70f && !Goals.HasActiveGoal(GoalType.FindWater))
                Goals.AddGoal(new Goal(GoalType.FindWater, "I need water", Vitals.Thirst, expiry));

            if (Vitals.Energy < 0.30f && !Goals.HasActiveGoal(GoalType.Rest))
                Goals.AddGoal(new Goal(GoalType.Rest, "I need rest", 1f - Vitals.Energy, expiry));
        }

        // ─── Lifecycle ────────────────────────────────────────────────────────────
        public void Deactivate() => IsActive = false;

        public override string ToString() =>
            $"[NPC] {Identity.DisplayName} (ID:{Identity.NpcId}) | {(IsActive ? "Active" : "Inactive")} | {Vitals} | {Psychology} | {Inventory}";
    }
}
